File: src/commonMain/kotlin/com/scoreo/domain/model/GameType.kt
package com.scoreo.domain.model

import kotlinx.serialization.Serializable

@Serializable
data class GameType(
    val id: String,
    val name: String,
    val winCondition: WinCondition,
    val tieBreakRule: TieBreakRule = TieBreakRule.NONE,
    val tieBreakCondition: WinCondition = WinCondition.HIGHEST_SCORE,
    val tieBreakLabel: String? = null,
    val active: Boolean = true,
) {
    fun computeWinners(playerScores: List<PlayerScore>): List<String> =
        computeWinners(playerScores, winCondition)

    fun computeWinners(playerScores: List<PlayerScore>, condition: WinCondition): List<String> {
        if (playerScores.isEmpty()) return emptyList()
        return when (condition) {
            WinCondition.HIGHEST_SCORE -> {
                val max = playerScores.maxOf { it.score }
                playerScores.filter { it.score == max }.map { it.playerId }
            }
            WinCondition.LOWEST_SCORE -> {
                val min = playerScores.minOf { it.score }
                playerScores.filter { it.score == min }.map { it.playerId }
            }
            WinCondition.MANUAL -> emptyList()
        }
    }
}

File: src/domain/model/gameType.ts
import type { TieBreakRule, WinCondition } from './enums'
import type { PlayerScore } from './playerScore'

export interface GameType {
  id: string
  name: string
  winCondition: WinCondition
  tieBreakRule: TieBreakRule
  tieBreakCondition: WinCondition
  tieBreakLabel: string | null
  active: boolean
}

export function computeWinners(
  gameType: GameType,
  playerScores: PlayerScore[],
  condition: WinCondition = gameType.winCondition,
): string[] {
  if (playerScores.length === 0) return []
  switch (condition) {
    case 'HIGHEST_SCORE': {
      const max = Math.max(...playerScores.map((s) => s.score))
      return playerScores.filter((s) => s.score === max).map((s) => s.playerId)
    }
    case 'LOWEST_SCORE': {
      const min = Math.min(...playerScores.map((s) => s.score))
      return playerScores.filter((s) => s.score === min).map((s) => s.playerId)
    }
    case 'MANUAL':
      return []
  }
}
